crates/bevy_animation/src/lib.rs

```diff
@@ -397,196 +493,226 @@ struct AnimationTargetContext<'a> {
+                ref blend_weights,
+                ref mut schedule,
+                ..
+            } = *player;
 
-        // Advance transition animations.
-        player.transitions.retain_mut(|transition| {
-            // Decrease weight. Expire the transition if necessary.
-            transition.current_weight -= transition.weight_decline_per_sec * time.delta_seconds();
-            if transition.current_weight <= 0.0 {
-                return false;
-            }
+            // Tick animations, and schedule them.
 
-            if let Some(animation_clip) = animation_clips.get(&transition.animation.animation_clip)
-            {
-                transition
-                    .animation
-                    .update(time.delta_seconds(), animation_clip.duration);
-            };
+            schedule.clear();
 
-            true
+            let mut dfs = animation_graph.dfs();
+            let mut weights = HashMap::with_capacity(animation_graph.graph.node_count());
+
+            while let Some(node_index) = dfs.next(&animation_graph.graph) {
+                let node = &animation_graph[node_index];
+
+                let mut weight = node.weight;
+                for parent in animation_graph
+                    .graph
+                    .neighbors_directed(node_index, Direction::Incoming)
+                {
+                    if let Some(parent_weight) = weights.get(&parent) {
+                        weight *= parent_weight;
+                    }
+                }
+
+                if let Some(playing_animation) = playing_animations.get_mut(&node_index) {
+                    // Tick the animation if necessary.
+                    if !playing_animation.paused {
+                        if let Some(ref clip_handle) = node.clip {
+                            if let Some(clip) = animation_clips.get(clip_handle) {
+                                playing_animation.update(delta_seconds, clip.duration);
+                            }
+                        }
+                    }
+
+                    weight *= playing_animation.weight;
+                } else if let Some(&blend_weight) = blend_weights.get(&node_index) {
+                    weight *= blend_weight;
+                }
+
+                weights.insert(node_index, weight);
+
+                if let (Some(clip), Some(playing_animation)) =
+                    (&node.clip, playing_animations.get(&node_index))
+                {
+                    schedule.push(ScheduledAnimationClip {
+                        clip: clip.clone(),
+                        seek_time: playing_animation.seek_time,
+                        weight,
+                    });
+                }
+            }
+
+            // Sort animations by weight, so that the most heavily-weighted
+            // animations come first.
+            //
+            // FIXME: This seems dubious, but it matches Bevy 0.13 behavior. Figure
+            // out what we want to do here.
```

Comment:
IMO we should omit this given the non-commutative nature of some of the blendings and explicitly mark this as changed behavior. It also tentatively breaks determinism if the animation movements are gameplay impacting.

Reply:
See my other comment: normalization broke transitions.

The only time this causes a change in behavior vs. today is when you have two simultaneous transitions, and the second one to start is faster than the first. Today, the transition that happens first is always applied first. With my change, they switch orders once one weight passes the other. Replicating the behavior of today would require the `AnimationPlayer` to keep track of a new source of data, "transition ordering". That's contrary to the RFC as I understand it, which specifies that transitions are to be built as a layer *on top of* the graph.

Reply:
I fixed this. See the new comment that exhaustively explains the math behind it.